sqltest/generic.go

```diff
@@ -73,8 +73,19 @@ func mustTruncateTables(t testing.TB, tr Truncator, tables ...string) {
 func truncateTables(t testing.TB, tr Truncator, agent Agent, tables ...string) error {
 	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
 	defer cancel()
+	ch := make(chan error)
 	for _, table := range tables {
-		if _, err := agent.ExecContext(ctx, fmt.Sprintf(truncateStmtFmt, table)); err != nil {
+		go func(table string) {
+			if _, err := agent.ExecContext(ctx, fmt.Sprintf(truncateStmtFmt, table)); err != nil {
+				ch <- err
```

Comment:
Would this block if there's an error in the channel already?

Reply:
It would, that is part of the synchronisation. The db query can run, but not pass the error back until the channel is clear.

Though what happens to those blocked goroutines when a faster goroutine returns an error and the main thread returns? I should prob kill them

Reply:
Sending anything to channels should be non-blocking imo, otherwise we're creating a semaphore for error reporting. We're also never draining the entire channel. So as soon as an error is found, we're blocking any further errors. I think the range should be on the error channel regardless and for the channel to close after the go-routines have finished to make sure it unlocks.

https://play.golang.org/p/Q7SXlDoMCCK

```go
package main

import (
	"fmt"
	"sync"
)

func main() {
	errC := make(chan error)
	wg := &sync.WaitGroup{}
	wg.Add(10)
	for i := 1; i <= 10; i++ {
		go func(i int) {
			defer wg.Done()
			fmt.Printf("process: %d\n", i)
			if i == 4 || i == 5 {
				errC <- fmt.Errorf("process error: %d", i)
			}
		}(i)
	}
	go func() {
		wg.Wait()
		close(errC)
	}()
	for err := range errC {
		fmt.Println(err)
		return
	}
}
```

Reply:
This will still block on operations that are taking too long.

Reply:
as soon as an error is found, the whole thing returns. I intended to only return the first error encountered. You are right that the other goroutines will be blocked after that first error returns, so I've updated the change to cancel those blocked goroutines

Reply:
Cool, cheers.